framework/src/main/java/org/checkerframework/framework/util/typeinference8/constraint/TypeConstraint.java

```diff
@@ -27,14 +28,65 @@ public abstract class TypeConstraint implements Constraint {
   /** T, the type on the right hand side of the constraint; may contain inference variables. */
   protected AbstractType T;
 
+  /** The constraint whose reduction created this constraint. */
+  public @Nullable Constraint parent;
+
+  /** A string that describes where this constraint is from. */
+  public @Nullable String source;
+
+  /**
+   * Creates a type constraint
+   *
+   * @param source a string describing where this constraint came from
+   * @param T the type of the right hand side of the constraint
+   */
+  protected TypeConstraint(String source, AbstractType T) {
+    assert T != null : "Can't create a constraint with a null type.";
+    this.T = T;
+    this.parent = null;
+    this.source = source;
+  }
+
   /**
    * Creates a type constraint
    *
+   * @param parent the constraint whose reduction created this constraint
    * @param T the type of the right hand side of the constraint
    */
-  protected TypeConstraint(AbstractType T) {
+  protected TypeConstraint(Constraint parent, AbstractType T) {
     assert T != null : "Can't create a constraint with a null type.";
     this.T = T;
+    this.parent = parent;
+    this.source = null;
+  }
+
+  /**
+   * Returns a string explain where this constraint came from.
+   *
+   * @return a string explain where this constraint came from
+   */
+  public String constraintHistory() {
+    StringBuilder constraintStack = new StringBuilder(this.toString()).append("\n");
```

Comment:
I think it would be a bit cleaner to use StringJoiner here.
also, please don't use "\n" in Java code.  Use System.lineSeparator() instead.